Enable use of epoch datetime representation in ObsSpace

srherbener opened the pull request:
## Description

This PR enables the use of the new epoch style representation for datetime values. This is being done to address performance issues related to using ISO 8601 formatted strings for datetime values.

This upgrade was done in a backward compatible fashion so current ioda files should continue to work without changes. However, the performance speedup will not be realized until the ioda file has its datetime values converted to the epoch representation.

### Issue(s) addressed

See jcsda-internal/ioda/pull/525 for details on what this PR helps fix.

## Acceptance Criteria (Definition of Done)

- All ioda, ufo and fv3-jedi tests pass
- Significant speed up of known problematic cases
    - Start with satwind file case from NCAR

## Dependencies

- [ ] merge with jcsda-internal/oops/pull/1550
- [ ] merge with jcsda-internal/ioda/pull/525
- [ ] merge with jcsda-internal/ufo/pull/1683
- [ ] merge with jcsda-internal/ioda-data/pull/57
- [ ] merge with jcsda-internal/um-jedi/pull/348

## Impact

None

srherbener committed:
Changed from string datetime (MetaData/datetime) to epoch datetime (MetaData/dateTime).

srherbener committed:
Merge branch 'develop' into feature/epoch-datetimes

ctgh commented:
To help my understanding, can you please explain why dateTime:units is equal to "seconds since 2019-01-01T00:00:00Z" in testinput_tier_1/met_office_profile_horizontal_drift.nc4?

SusanSun17 committed:
Change report flags for derived obs value

srherbener commented:
> To help my understanding, can you please explain why dateTime:units is equal to "seconds since 2019-01-01T00:00:00Z" in testinput_tier_1/met_office_profile_horizontal_drift.nc4?

@ctgh the new epoch style representation is implemented with a single variable where the units attribute defines the epoch and the variable itself contains int64_t values that are offsets in seconds. The "seconds since 2019-01-01T00:00:00Z" was an arbitrary choice of using Jan 1, 2019 at 00Z for the epoch spec. The epoch can be any datetime. The format of the units is from netcdf-CF and COARDS conventions. One side benefit is that ncdump recognizes this specification of units and when you use the -t option of ncdump, it will transform the datetime values to human readable form.

Does this answer your question?

ctgh commented:
@srherbener Yes, that's very useful - thanks!

srherbener committed:
Workaround fix for the test_ufo_obserror_assign_unittests ctest failures. The DataExtractor needs work to
process the new epoch style datetimes and we decided to defer that work until later.

srherbener committed:
Merge branch 'develop' into feature/epoch-datetimes

srherbener committed:
Merge branch 'develop' into feature/epoch-datetimes

srherbener committed:
Merge branch 'develop' into feature/epoch-datetimes

ctgh committed:
updated test file to use dateTime
